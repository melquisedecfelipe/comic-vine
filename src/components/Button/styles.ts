import styled, { css, DefaultTheme } from 'styled-components'

import { ButtonProps } from '.'

export type WrapperProps = Pick<ButtonProps, 'variant'>

const wrapperModifiers = {
  filled: (theme: DefaultTheme) => css`
    color: ${theme.colors.white};
    background: ${theme.colors.black};
  `,
  outlined: (theme: DefaultTheme) => css`
    color: ${theme.colors.black};
    background: ${theme.colors.white};
    border: 1px solid ${theme.colors.black};

    &:hover {
      color: ${theme.colors.white};
      background: ${theme.colors.black};
    }
  `,
  disabled: (theme: DefaultTheme) => css`
    &:disabled {
      cursor: not-allowed;
      background: ${theme.colors.disable};
      color: ${theme.colors.grey};
      border: 0;
    }
  `
}

export const Wrapper = styled.button<WrapperProps>`
  ${({ theme, variant, disabled }) => css`
    display: inline-flex;
    align-items: center;
    justify-content: center;
    font-size: ${theme.font.familyBarlow};
    font-weight: ${theme.font.semiBold};
    font-size: ${theme.font.sizes.medium};
    padding: ${theme.spacings.small} ${theme.spacings.medium};
    cursor: pointer;
    border: none;
    text-decoration: none;
    transition: ${theme.transition.default};
    height: ${theme.font.sizes.huge};

    &:focus {
      opacity: 0.75;
    }

    ${!!variant && wrapperModifiers[variant](theme)};
    ${disabled && wrapperModifiers.disabled(theme)};
  `}
`
