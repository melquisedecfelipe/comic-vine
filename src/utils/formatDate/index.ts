export const formatDate = (date?: string | null): string => {
  if (!date) {
    return ''
  }

  const newDate = new Date(date)

  return new Intl.DateTimeFormat('en-US', {
    day: 'numeric',
    month: 'short',
    year: 'numeric'
  }).format(newDate)
}

export const formatDateToIso = (date?: string | null): string => {
  if (!date) {
    return ''
  }

  const newDate = new Date(date)

  return newDate.toISOString().substring(0, 10)
}
