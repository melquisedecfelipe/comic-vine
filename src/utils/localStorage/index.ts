import { Character, CharacterEdit } from 'types/characters'

export function getStorageItem(key: string) {
  if (typeof window === 'undefined') return

  const data = window.localStorage.getItem(key)

  return JSON.parse(data!)
}

export function setStorageItem(
  key: string,
  value: string[] | Character[] | CharacterEdit
) {
  if (typeof window === 'undefined') return

  const data = JSON.stringify(value)

  return window.localStorage.setItem(key, data)
}
