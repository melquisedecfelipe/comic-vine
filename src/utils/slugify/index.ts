export const slugify = (id: number, slug: string): string => {
  const string = slug.split(' ').join('-').toLocaleLowerCase()

  return [id, string].join('-')
}
