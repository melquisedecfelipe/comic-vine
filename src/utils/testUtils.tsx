import { ReactElement } from 'react'

import { render, RenderOptions } from '@testing-library/react'

import { ThemeProvider } from 'styled-components'

import {
  CharactersContext,
  CharactersContextData,
  CharactersContextDefaultValues
} from 'hooks/useCharacters'

import theme from 'styles/theme'

type CustomRenderProps = {
  charactersProviderProps?: CharactersContextData
} & Omit<RenderOptions, 'queries'>

const customRender = (
  ui: ReactElement,
  {
    charactersProviderProps = CharactersContextDefaultValues,
    ...renderOptions
  }: CustomRenderProps = {}
) =>
  render(
    <ThemeProvider theme={theme}>
      <CharactersContext.Provider value={charactersProviderProps}>
        {ui}
      </CharactersContext.Provider>
    </ThemeProvider>,
    renderOptions
  )

export * from '@testing-library/react'
export { customRender as render }
