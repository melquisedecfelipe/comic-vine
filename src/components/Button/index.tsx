import { forwardRef, AnchorHTMLAttributes, ButtonHTMLAttributes } from 'react'

import * as S from './styles'

type ButtonTypes =
  | AnchorHTMLAttributes<HTMLAnchorElement>
  | ButtonHTMLAttributes<HTMLButtonElement>

export type ButtonProps = {
  as?: React.ElementType
  variant?: 'filled' | 'outlined'
} & ButtonTypes

const Button: React.ForwardRefRenderFunction<S.WrapperProps, ButtonProps> = (
  { children, variant = 'filled', ...props },
  ref
) => (
  <S.Wrapper variant={variant} ref={ref} {...props}>
    {children}
  </S.Wrapper>
)

export default forwardRef(Button)
