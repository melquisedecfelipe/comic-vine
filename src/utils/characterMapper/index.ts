import { slugify } from 'utils/slugify'

import { BBFCharacter, Character, CharacterDetail } from 'types/characters'

const genders = new Map()

genders.set(0, 'Unknown')
genders.set(1, 'Male')
genders.set(2, 'Female')

export const characterMapper = (character: BBFCharacter): CharacterDetail => {
  return {
    aliases: character.aliases,
    detail: character.api_detail_url,
    birth: character.birth,
    gender: genders.get(character.gender),
    id: character.id,
    images: {
      icon: character.image.icon_url,
      screen: character.image.screen_url,
      small: character.image.small_url
    },
    name: character.name,
    realName: character.real_name,
    deck: character.deck,
    description: character.description,
    origin: character.origin.name,
    publisher: character.publisher.name,
    site: character.site_detail_url,
    slug: slugify(character.id, character.name)
  }
}

export const charactersMapper = (characters: BBFCharacter[]): Character[] => {
  return characters.map(character => ({
    aliases: character.aliases,
    detail: character.api_detail_url,
    birth: character.birth,
    id: character.id,
    gender: genders.get(character.gender),
    images: {
      icon: character.image.icon_url,
      screen: character.image.screen_url,
      small: character.image.small_url
    },
    name: character.name,
    realName: character.real_name,
    deck: character.deck,
    description: character.description,
    site: character.site_detail_url,
    slug: slugify(character.id, character.name)
  }))
}
