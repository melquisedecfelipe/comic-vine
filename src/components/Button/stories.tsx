import { Story, Meta } from '@storybook/react/types-6-0'

import Button from '.'

export default {
  title: 'Button',
  component: Button,
  args: {
    children: 'Button Component'
  },
  argTypes: {
    children: {
      type: 'string'
    },
    onClick: { action: 'buttonClick' }
  }
} as Meta

export const Default: Story = args => <Button {...args} />

export const asLink: Story = args => <Button {...args} />

asLink.args = {
  variant: 'outlined',
  children: 'Link Component',
  as: 'a'
}

asLink.argTypes = {
  onClick: { action: 'linkClick' }
}
