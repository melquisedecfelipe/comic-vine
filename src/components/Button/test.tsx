import { render, screen } from 'utils/testUtils'

import Button from '.'

describe('<Button />', () => {
  it('should render the filled style by default', () => {
    const { container } = render(<Button>Button component</Button>)

    expect(
      screen.getByRole('button', { name: /button component/i })
    ).toHaveStyle({
      padding: '0px 16px',
      'font-size': '16px',
      color: '#FFFFFF',
      background: '#000000'
    })

    expect(container.firstChild).toMatchSnapshot()
  })

  it('should render the outlined style', () => {
    render(<Button variant="outlined">Button component</Button>)

    expect(
      screen.getByRole('button', { name: /button component/i })
    ).toHaveStyle({
      color: '#000000',
      background: '#FFFFFF',
      border: '1px solid #000000'
    })
  })

  it('should render a disabled Button', () => {
    render(<Button disabled>Button component</Button>)

    expect(
      screen.getByRole('button', { name: /button component/i })
    ).toHaveStyleRule('cursor', 'not-allowed', {
      modifier: ':disabled'
    })
  })

  it('should render Button as a link', () => {
    render(
      <Button as="a" href="/link">
        Button as link
      </Button>
    )

    expect(
      screen.getByRole('link', { name: /button as link/i })
    ).toHaveAttribute('href', '/link')
  })
})
