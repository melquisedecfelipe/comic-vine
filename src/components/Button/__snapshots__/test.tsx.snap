// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Button /> should render the filled style by default 1`] = `
.c0 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  font-size: Barlow,-apple-system,BlinkMacSystemFont,'Segoe UI',Roboto,Oxygen,Ubuntu,Cantarell,'Open Sans','Helvetica Neue',sans-serif;
  font-weight: 600;
  font-size: 16px;
  padding: 0 16px;
  cursor: pointer;
  border: none;
  -webkit-text-decoration: none;
  text-decoration: none;
  -webkit-transition: 0.3s ease-in-out;
  transition: 0.3s ease-in-out;
  min-height: 40px;
  max-height: 40px;
  color: #FFFFFF;
  background: #000000;
}

.c0:focus {
  opacity: 0.75;
}

<button
  class="c0"
>
  Button component
</button>
`;
